pycroft/model/traffic.py

```diff
@@ -222,5 +225,146 @@ class CurrentTrafficBalance(ModelBase):
+            else_=None)).label(label)
+
+
+    hist = select([buckets.c.bucket,
+                   cond_sum(events.c.type == 'Credit', 'credit'),
+                   cond_sum(events.c.type == 'Ingress', 'ingress', invert=True),
+                   cond_sum(events.c.type == 'Egress', 'egress', invert=True),
+                   func.sum(events.c.amount).label('amount'),
+                   cond_sum(and_(balance_timestamp != None, events.c.timestamp < balance_timestamp), 'before_balance'),
+                   cond_sum(or_(balance_timestamp == None, events.c.timestamp >= balance_timestamp), 'after_balance')]
+    ).select_from(buckets.outerjoin(
+        events, func.width_bucket(
+            events.c.timestamp, select([func.array(select([buckets.c.bucket]).select_from(buckets).where(buckets.c.index != 0).label('dummy'))])
+        ) == buckets.c.index
+    )).where(
+        # Discard bucket n+1
+        buckets.c.index < select([func.max(buckets.c.index)])
+    ).group_by(
+        buckets.c.bucket
+    ).order_by(
+        buckets.c.bucket
+    ).cte('traffic_hist')
+
+
+    # Bucket is located before the balance and no traffic_events exist before it
+    first_event_timestamp = select([func.min(events.c.timestamp)]).as_scalar()
+    case_before_balance_no_data = (
+        and_(balance_timestamp != None, hist.c.bucket < balance_timestamp,
+        or_(first_event_timestamp == None,
+            hist.c.bucket < first_event_timestamp
+            )),
+        None
+    )
+
+    # Bucket is located after the balance
+    case_after_balance = (
+        or_(balance_timestamp == None, hist.c.bucket >= balance_timestamp),
+        func.coalesce(balance_amount, 0) + func.coalesce(
+            func.sum(hist.c.after_balance).over(
+                order_by=hist.c.bucket.asc(), rows=(None, 0)),
+            0)
+    )
+
+    # Bucket is located before the balance, but there still exist traffic_events before it
+    else_before_balance = (
+            func.coalesce(balance_amount, 0) +
+            func.coalesce(hist.c.after_balance, 0) -
+            func.coalesce(
+                func.sum(hist.c.before_balance).over(
+                    order_by=hist.c.bucket.desc(), rows=(None, -1)
+                ), 0)
+    )
+
+    agg_hist = select(
+            [hist.c.bucket, hist.c.credit, hist.c.ingress, hist.c.egress, case(
+            [case_before_balance_no_data, case_after_balance],
+            else_=else_before_balance
+        ).label('balance')]).alias('agg_hist')
+
+    # Remove bucket 0
+    result = select([agg_hist]).order_by(agg_hist.c.bucket).offset(1)
+
+    return result
+
+
+traffic_history_function = Function(
+    'traffic_history', ['arg_user_id int', 'arg_start timestamptz', 'arg_interval interval', 'arg_step interval'],
+    'TABLE ("timestamp" timestamptz, credit numeric, ingress numeric, egress numeric, balance numeric)',
+    str(
+        traffic_history_query().compile(
```

Comment:
OK, I am just gonna trust this one.
ph'nglui mglw'nafh Cthulhu R'lyeh wgah'nagl fhtagn. Please don't let me down.

Reply:
a hideous function like that requires a docstring though 😕

Reply:
@mueslo no it doesn't. Be quiet.